### PR TITLE
Danaswap UI

### PR DESCRIPTION
**Pull request by leomayleomay**
Closes

This PR is to setup infrastructure for UI

other changes

- [ ] The impure tests via `nix run .#offchain:test` were run and all are passing.
- [ ] Where applicable, grammar, punctuation, and spelling within code and comments should be plausibly correct.
- [ ] The issue(s) that the work is addressing should be linked in the PR description.
- [ ] Any significant changes beyond what's clear from the linked issue(s) are documented in the PR description

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "build": "node build.js"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "@esbuilder/html": "^0.0.5",
+    "autoprefixer": "^10.4.4",
+    "ejs": "^3.1.6",
+    "esbuild": "^0.14.36",
```
Comment (marijanp) [inline ui/package.json:15]: Just pin it to `0.14.38`

File: ui/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "danaswap-ui",
+  "version": "1.0.0",
+  "description": "",
+  "scripts": {
+    "build": "node build.js"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
```
Comment (marijanp) [inline ui/package.json:11]: shouldnt these be `dependencies`

Comment (leomayleomay) [reply ui/package.json:11]: these deps are here for building the static HTML/CSS, they will not be used in production

**Comment by leomayleomay** [comment]
Deployed to Netlify:
* https://63506e08cee2e2105dd41805--danaswap-ui.netlify.app/
* https://63506e08cee2e2105dd41805--danaswap-ui.netlify.app/pools

**Review by marijanp — CHANGES_REQUESTED** [review]
Is the file `ui/assets/css/home.css` intentionally empty?

File: ui/flake-module.nix [hunk 3]
```diff
@@ -0,0 +1,83 @@
+{ self, ... }:
+{
+  perSystem = { config, self', inputs', system, ... }:
+    let
+      pkgs = inputs'.nixpkgs.legacyPackages;
+      inherit (config) cat-lib dream2nix;
+      ui =
+        (dream2nix.lib.makeOutputs { source = ./.; settings = [{ subsystemInfo.nodejs = 16; }]; }).packages.danaswap-ui;
```
[the remark this answers is not in the document]
Comment (leomayleomay) [reply ui/flake-module.nix:8]: I try to keep danaswap-browser and UI separated, the browser is to provide components (written in Purescript, built with purs-nix) that will be used in UI (written in EJS, built with dream2nix + build.js)

[the remark this answers is not in the document]
Comment (marijanp) [reply ui/flake-module.nix:8]: Yes, but you don't need to write a build script in javascript in order to bundle the purescript code with the javascript components. So that is why I'm asking

[the remark this answers is not in the document]
Comment (marijanp) [reply ui/flake-module.nix:8]: Okay I see now what you did, had to look a bit closer. So you build both things independently and then you just copy files around to the right places

File: ui/flake-module.nix [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    in
+    {
+      packages = {
+        danaswap-ui =
+          pkgs.runCommand "build-danaswap-ui"
+            { buildInputs = [ self'.packages."offchain:danaswap-browser" ]; }
```
Comment (marijanp) [inline ui/flake-module.nix:14]:
```suggestion
            { }
```
not needed as you directly refer to the package using interpolation below

File: offchain/webpack.config.js [hunk 5]
```diff
@@ -76,11 +76,6 @@ module.exports = {
     new webpack.LoaderOptionsPlugin({
       debug: true,
     }),
-    new HtmlWebpackPlugin({
-      title: "hello-world-api",
-      template: "./index.html",
-      inject: false, // See stackoverflow.com/a/38292765/3067181
-    }),
```
Comment (marijanp) [inline offchain/webpack.config.js:83]: Why was this deleted? this `webpack.config.js` was copied from https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/webpack.config.js

File: offchain/danaswap-browser/flake-module.nix [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+              cp -r ${danaswap-browser.ps.modules."DanaSwapBrowser.Home".output { }} homeOutput
+              cp -r ${danaswap-browser.ps.modules."DanaSwapBrowser.Pools".output { }} poolsOutput
+              cp ${./home.js} home.js
+              cp ${./pools.js} pools.js
+              cp -r ${nodeModules}/* .
+              export NODE_PATH="node_modules"
+              export PATH="bin:$PATH"
+              mkdir -p $out/dist
+              webpack --mode=production -c ${../webpack.config.js} -o $out/dist/home --entry ./home.js
+              webpack --mode=production -c ${../webpack.config.js} -o $out/dist/pools --entry ./pools.js
```
Comment (marijanp) [inline offchain/danaswap-browser/flake-module.nix:48]: Do you think it could make sense to make the `pool` and `home` separate package outputs?

Comment (marijanp) [reply offchain/danaswap-browser/flake-module.nix:48]: That way we could factor out one function which we can use to bundle both pool and home

**Review by marijanp — CHANGES_REQUESTED** [review]
LGTM, I was a little bit confused looking at this the first time but now I understand what you did.

I have a few questions though.
Can you explain where `ui/assts` is used? the files in there are empty and I can't see any usage.

**Comment by leomayleomay** [comment]
> Is the file `ui/assets/css/home.css` intentionally empty?

Indeed, I will delete it since it's not used

File: offchain/webpack.config.js (hunk 5 above)
Comment (leomayleomay) [reply offchain/webpack.config.js:83]: yes, it was copied from CTL then I found the stanza is useless since we're not hosting an index page as CTL does

File: offchain/danaswap-browser/src/Home.purs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+module DanaSwapBrowser.Home where
```
[the remark this answers is not in the document]
Comment (leomayleomay) [reply offchain/danaswap-browser/src/Home.purs:1]: sure thing, will do

File: offchain/danaswap-browser/flake-module.nix (hunk 6 above)
Comment (leomayleomay) [reply offchain/danaswap-browser/flake-module.nix:48]: > Do you think it could make sense to make the `pool` and `home` separate package outputs?

yes, they will be imported by UI separately, hence separate package outputs

Comment (leomayleomay) [reply offchain/danaswap-browser/flake-module.nix:48]: > That way we could factor out one function which we can use to bundle both pool and home

@marijanp I did try to introduce a function `mkBundle`, 

```
      let
         mkBundle = module: outputDir: entryJs: distDir:
              let
                moduleFullPath = danaswap-browser.ps.modules."${module}".output { };
              in
              ''
                export BROWSER_RUNTIME=1
                cp -r ${moduleFullPath} ${outputDir}
                cp ${./${entryJs}} ${entryJs}
                cp -rp ${nodeModules}/* .
                export NODE_PATH="node_modules"
                export PATH="bin:$PATH"
                mkdir -p $out/dist
                webpack --mode=production -c ${../webpack.config.js} -o $out/dist/${distDir} --entry ./${entryJs}
              '';
            homeBundle = mkBundle "DanaSwap.Browser.Home" "homeOutput" "home.js" "home";
            poolsBundle = mkBundle "DanaSwap.Browser.Pools" "poolsOutput" "pools.js" "pools";
          in
          pkgs.runCommand "build-danaswap-browser" { } (lib.concatStrings [homeBundle poolsBundle]);
```

it will raise an error since copying `node_modules` to the same destination

```
error: builder for '/nix/store/50kg1pgfcfg4gcvhfbkz6f7iqc4ysld8-build-danaswap-browser.drv' failed with exit code 1;
       last 10 log lines:
       > cp: cannot remove './node_modules/assert/LICENSE': Permission denied
       > cp: cannot remove './node_modules/assert/build/assert.js': Permission denied
       > cp: cannot remove './node_modules/assert/build/internal/util/comparisons.js': Permission denied
       > cp: cannot remove './node_modules/assert/build/internal/errors.js': Permission denied
       > cp: cannot remove './node_modules/assert/build/internal/assert/assertion_error.js': Permission denied
       > cp: cannot remove './node_modules/parseurl/HISTORY.md': Permission denied
       > cp: cannot remove './node_modules/parseurl/package.json': Permission denied
       > cp: cannot remove './node_modules/parseurl/README.md': Permission denied
       > cp: cannot remove './node_modules/parseurl/LICENSE': Permission denied
       > cp: cannot remove './node_modules/parseurl/index.js': Permission denied
       For full logs, run 'nix log /nix/store/50kg1pgfcfg4gcvhfbkz6f7iqc4ysld8-build-danaswap-browser.drv'.
```

Comment (marijanp) [reply offchain/danaswap-browser/flake-module.nix:48]: yeah the problem is that you concat the string and this means that the script copies the node-modules twice which is in itself a little bit ugly. I'ld say lets leave it like that for now

File: offchain/danaswap-browser/src/Home.purs (hunk 7 above)
[the remark this answers is not in the document]
Comment (marijanp) [reply offchain/danaswap-browser/src/Home.purs]: You didn't move the files

**Review by marijanp — CHANGES_REQUESTED** [review]
Move the files in `danaswap-browser` in the meantime I'm looking into the not terminating effect

File: offchain/danaswap-browser/src/Home.purs (hunk 7 above)
[the remark this answers is not in the document]
Comment (leomayleomay) [reply offchain/danaswap-browser/src/Home.purs]: ahh, I renamed the modules but forgot to move them into sub folder, will do it now

**Review by marijanp — APPROVED** [review]
Will merge once pipeline succeeds. Great job!
